bal.1
.\" Copyright (C) Zach Flynn <zlflynn@gmail.com>, 2018  
.\" You may distribute this file under the terms of the FreeBSD Documentation License which is the text below:
\" FreeBSD Documentation License
\" Copyright 2018 Zach Flynn. All rights reserved.

\" Redistribution and use in source (Groff) and 'compiled' forms (SGML, HTML, PDF, PostScript, RTF and so forth) with or without modification, are permitted provided that the following conditions are met:

\" Redistributions of source code (Groff) must retain the above copyright notice, this list of conditions and the following disclaimer as the first lines of this file unmodified.

\" Redistributions in compiled form (transformed to other DTDs, converted to PDF, PostScript, RTF and other formats) must reproduce the above copyright notice, this list of conditions and the following disclaimer in the documentation and/or other materials provided with the distribution.

\" THIS DOCUMENTATION IS PROVIDED BY ZACH FLYNN "AS IS" AND ANY EXPRESS OR IMPLIED WARRANTIES, INCLUDING, BUT NOT LIMITED TO, THE IMPLIED WARRANTIES OF MERCHANTABILITY AND FITNESS FOR A PARTICULAR PURPOSE ARE DISCLAIMED. IN NO EVENT SHALL ZACH FLYNN BE LIABLE FOR ANY DIRECT, INDIRECT, INCIDENTAL, SPECIAL, EXEMPLARY, OR CONSEQUENTIAL DAMAGES (INCLUDING, BUT NOT LIMITED TO, PROCUREMENT OF SUBSTITUTE GOODS OR SERVICES; LOSS OF USE, DATA, OR PROFITS; OR BUSINESS INTERRUPTION) HOWEVER CAUSED AND ON ANY THEORY OF LIABILITY, WHETHER IN CONTRACT, STRICT LIABILITY, OR TORT (INCLUDING NEGLIGENCE OR OTHERWISE) ARISING IN ANY WAY OUT OF THE USE OF THIS DOCUMENTATION, EVEN IF ADVISED OF THE POSSIBILITY OF SUCH DAMAGE.

.TH bal 1 2019-04-26 
.SH NAME
bal \- for keeping accounts in order and studying past spending habits
.SH SYNOPSIS
.B bal [-f FILE] [-l CODE] [-s] [-n] [COMMAND]

.SH DESCRIPTION
.B bal
keeps track of personal accounts and spending.  

.LP
Text entry at the
.B bal
console follows standard line editor conventions (it has Emacs-like
key bindings, uses the GNU readline library). To cancel a
command, type: Control-C.


.B bal's
main interactive commands are:

.TP
.BR aa
add account
.TP
.BR at
add transaction to current account
.TP
.BR et\~ [n]
edit transaction, optionally setting how many of the latest transaction numbers to list (any transaction can be chosen whether it is listed or not).
.TP
.BR lt
list most recent transactions in current account (change number to list by modifying the variable bal/number-to-quick-list).
.TP
.BR ltn
list most recent transactions in current account, prompt for number to
list
.TP
.BR ea
rename account
.TP
.BR da
delete account
.TP
.BR dt
delete transaction
.TP
.BR la
list accounts and current balances
.TP
.BR lae
list expense accounts and current balances
.TP
.BR lai
list income accounts and current balances
.TP
.BR laa
list asset accounts and current balances
.TP
.BR lal
list liability accounts and current balances
.TP
.BR bt
total accounts by type (income, expense, asset, liability)
.TP
.BR cex
total across expenditure accounts
.TP
.BR cin
total across income accounts
.TP
.BR cas
total across asset accounts
.TP
.BR cli
total across liability accounts
.TP
.BR cwo
assets - liabilities
.TP
.BR cba
current balance
.TP
.BR re
list transactions in current account matching a regular expression
.TP
.BR sa
set current account
.TP
.BR ca
display current account
.TP
.BR w
write accounts to file
.TP
.BR r
read saved accounts from file
.TP
.BR t
transfer from one account to another
.TP
.BR dtr
delete a transfer
.TP
.BR ltbd
list transactions in current account by day, prompting for the first and last day to show
.TP
.BR v
print version
.TP
.BR sd
set "current" day, changing it allows you to see how account values have changed over time 
.TP
.BR cd
print current day out in YYYY-MM-DD format
.TP
.BR baod
prints account balance over a sequence of days
.TP
.BR exod
prints total expenses as of a sequence of days
.TP
.BR inod
prints total income as of a sequence of days
.TP
.BR liod
prints total liabilities as of a sequence of days
.TP
.BR wood
prints net worth as of a sequence of days
.TP
.BR ttbd
prints total of all transactions between a sequence of days
.TP
.BR ttre
prints total of all transactions between a sequence of days matching a regular expression
.TP
.BR pl
pay a loan, allows you to split a payment between a principal account (liability account) and an interest account (an expense account)
.TP
.BR fn
print out current default save file
.TP
.BR csp
Change the stock price of current account
.TP
.BR "cal [opt]"
print out calendar, can pass a string to cal which acts like specifying options for the program cal at the command line
.TP
.BR p " " X
print X, a Scheme expression
.TP
.BR q
quit, unless -s specified, prompt to save before quit
.LP
The command prompt is a full Guile Scheme interpreter except that you do not need to enclose the outer-most expression in parenthesis. Any valid Scheme code can be executed at the prompt.  For example, typing
.B display \(dqhello\(dq
will print the text "hello".
.B bal
loads the Scheme functions in the /etc/bal.scm file at startup. These functions implement the commands above.
.LP
Of particular interest for developing your own functions in Scheme is the function
.B bal/call
which can be used to make Scheme functions interactive.
.LP
The saved file is simply a
.B tar
archive of CSV files that record transactions and characteristics of accounts.  These files are human-readable.
.LP
The prompt is whatever is returned by the evaluating the Scheme expression "(bal/prompt)" so redefining this function will customize your prompt.
.SH OPTIONS
.TP
.BR \-f " " FILE
load accounts from FILE
.TP
.BR \-l " " CODE
execute Scheme code in file CODE
.TP
.BR \-s
do not prompt to save on exit, will not save on exit
.TP
.BR \-n
do not load ~/.balrc.scm
.TP
.BR COMMAND
execute COMMAND as if it was entered at
.B bal's
prompt and then exit.  This enables you to, for example, type "bal t" to enter a new transaction (supposing your main bal file is loaded in your ~/.balrc.scm).
.SH FILES
.B bal
executes Scheme code in ~/.balrc.scm if the file is present.
.SH EXAMPLES
.TP
bal
start the bal interpreter and execute any Scheme code in ~/.balrc.scm (potentially including code to load data from a default file).
.TP
bal -f myaccounts.btar
load the file 
.B myaccounts.btar
from the current directory.
.TP
bal -s v
quits and exists after displaying version number (without prompting to save data).
.TP
bal aa
displays prompts to add new account after evaluating ~/.balrc.scm and then exits (prompting to save)
.TP
bal t
displays prompts to add new transaction after evaluating ~/.balrc.scm and then exits (prompting to save)
.TP
bal -s la
displays current account balances and quits without prompting to save.
.SH "SEE ALSO"
.SH BUGS
.TP
Please report as an issue to https://github.com/flynnzac/bal
.SH AUTHOR
Zach Flynn <zlflynn@gmail.com>


bal.scm
(use-modules (ice-9 format))
(use-modules (srfi srfi-1))
(use-modules (srfi srfi-19))

;; parameters
(define bal/number-to-quick-list 20)
(define bal/prompt
  (lambda ()
    (if (= (bal/get-number-of-accounts) 0)
        ":> "
        (string-append
         "("
         (bal/get-current-account)
         ") :> "))))


;; non-interactive functions

;;; utility functions

(define bal/print-tscts
  (lambda (tsct-list)
    (if (list? tsct-list)
        (let ((len (apply max
                          (map (lambda (x)
                                 (string-length (car x))) tsct-list))))
          (map-in-order
           (lambda (x)
             (display
              (string-append
               (number->string (list-ref x 2))
               "-"
               (format #f "~2,'0d" (list-ref x 3))
               "-"
               (format #f "~2,'0d" (list-ref x 4))
               " "
               (format #f (string-append
                           "~"
                           (number->string (+ len 5))
                           "a") (list-ref x 0))
               " "
               (format #f "~10,2f" (list-ref x 1))
               "\n")))
           tsct-list)))))

(define bal/day-from-time
  (lambda (x)
    (let ((xdate (time-utc->date x 0)))
      (list (date-day xdate) (date-month xdate) (date-year xdate)))))

(define bal/seq-days
  (lambda (first-day last-day by)
    (let ((first-time (date->time-utc (make-date 0 0 0 0
                                                 (list-ref first-day 0)
                                                 (list-ref first-day 1)
                                                 (list-ref first-day 2)
                                                 0)))
          (last-time (date->time-utc (make-date 0 0 0 0
                                                (list-ref last-day 0)
                                                (list-ref last-day 1)
                                                (list-ref last-day 2)
                                                0))))
      (if (time>=? first-time last-time)
          (list first-day)
          (cons first-day (bal/seq-days 
                           (bal/day-from-time
                            (add-duration first-time
                                          (make-time time-duration 0 (* 24 3600 by))))
                           last-day
                           by))))))


;;; add and delete functions

(define bal/t
  (lambda (to-account from-account amount desc day)
    (bal/at to-account amount desc day)
    (bal/at from-account (* -1 amount) desc day)))

(define bal/dtr
  (lambda (from-account location)
    (let ((transaction
           (bal/get-transaction-by-location (car location)
                                            (cdr location)))
          (all-from (bal/get-all-transactions from-account))
          (num-from (bal/get-account-location from-account)))
      (bal/dt (cons num-from
                    (list-index
                     (lambda (u)
                       (and (string=? (car u) (car transaction))
                            (lset= (lambda (v w) (= (abs v) (abs w)))
                                   (cdr u) (cdr transaction)))) all-from)))
      (bal/dt location))))

(define bal/pay-loan
  (lambda (loan-account interest-account from-account
                        principal interest desc day)
    (bal/at loan-account principal desc day)
    (bal/at interest-account interest desc day)
    (bal/at from-account (* -1 (+ principal interest)) desc day)))

(define bal/change-stock-price
  (lambda (account from-account stock-price number day)
    (let* ((value (list-ref (bal/total-account account) 1))
           (newval (* number stock-price))
           (trval (- newval value)))
      (bal/t account from-account trval
             "Stock Price Change" day))))

;;; modification functions

(define bal/edit-transact
  (lambda (tsct-loc day amount desc)
    (let ((tsct (bal/get-transaction-by-location
                 (car tsct-loc) (cdr tsct-loc))))
      (bal/dt tsct-loc)
      (bal/at (car (bal/get-account-by-location (car tsct-loc)))
              (if (string-null? amount)
                  (list-ref tsct 1)
                  (string->number amount))
              (if (string-null? desc)
                  (list-ref tsct 0)
                  desc)
              day))))

;;; summary functions

(define bal/display-account-totals
  (lambda (accts)
    (let ((len
           (apply max
                  (map
                   (lambda (x) (string-length (car x))) accts))))
      (map-in-order
       (lambda (x)
         (display
          (string-append
           (format #f
                   (string-append "~"
                                  (number->string (+ len 5))
                                  "a") (car x))
           " "
           (format #f "~10,2f" (car (cdr x)))
           " "
           (format #f "~10,2f" (cdr (cdr x)))
           "\n")))
       accts))))           

(define bal/current-total-of-type
  (lambda (n)
    (lambda ()
      (let* ((accts (bal/total-by-account-type))
             (el (cdr (list-ref accts n))))
        (display (string-append
                  (format #f "~,2f"
                          (if (pair? el)
                              (car el)
                              el))
                  "\n"))))))

;;; over time functions

(define-syntax bal/loop-days
  (lambda (x)
    (syntax-case x ()
      ((_ days current-day val exp)
       (with-syntax ((i (datum->syntax x (quote i))))
         (syntax (let loop-day ((i 0))
                   (if (< i val)
                       (begin
                         (bal/set-current-day (list-ref days i))
                         (cons (cons (list-ref days i)
                                     exp)
                               (loop-day (+ i 1))))
                       (begin
                         (bal/set-current-day current-day)
                         (list))))))))))

(define bal/balance-account-on-days
  (lambda (first-day last-day by account)
    (let ((days (bal/seq-days first-day last-day by))
          (current-day (bal/get-current-day)))
      (bal/loop-days days current-day (length days)
                     (list-ref (bal/total-account account) 1)))))

(define bal/total-transact-in-account-between-days
  (lambda (first-day last-day by account)
    (let* ((balance (bal/balance-account-on-days
                     first-day last-day by account))
           (current-day (bal/get-current-day))
           (days (map car balance)))
      (bal/loop-days days current-day (- (length balance) 1)
                     (- (cdr (list-ref balance (+ i 1)))
                        (cdr (list-ref balance i)))))))

(define bal/total-transact-in-account-re
  (lambda (first-day last-day by account regex)
    (let ((current-day (bal/get-current-day))
          (days (bal/seq-days first-day last-day by)))
      (bal/loop-days days current-day (length days)
                     (apply +
                            (map (lambda (u) (list-ref u 1))
                                 (bal/get-transactions-by-regex account regex)))))))

(define bal/output-by-day
  (lambda (day amount)
    (display
     (string-append
      (number->string (list-ref day 2))
      "-"
      (format #f "~2,'0d" (list-ref day 1))
      "-"
      (format #f "~2,'0d" (list-ref day 0))
      " "
      (format #f "~10,2f" amount)
      "\n"))))

(define bal/get-by-type-over-days
  (lambda (first-day last-day by num)
    (let ((days (bal/seq-days first-day last-day by))
          (current-day (bal/get-current-day)))
      (bal/loop-days days current-day (length days)
                     (list-ref
                      (list-ref (bal/total-by-account-type) num)
                      1)))))

(define bal/get-by-type-over-days-for-type
  (lambda (n)
    (lambda (first-day last-day by)
      (if (string=? by "")
          (bal/get-by-type-over-days first-day last-day 7 n)
          (bal/get-by-type-over-days first-day last-day (string->number by) n)))))

(define-syntax bal/over-day-cmd
  (syntax-rules ()
    ((over-day-cmd val)
     (let ((result (bal/call
                    (string-append
                     "(bal/get-by-type-over-days-for-type "
                     (number->string val)
                     ")")
                    (list
                     (cons "First Day" "day")
                     (cons "Last Day" "day")
                     (cons "By [7]" "string")))))
       (map-in-order
        (lambda (x)
          (bal/output-by-day (car x) (cdr x)))
        result)))))



;; interactive functions

(define aa
  (lambda ()
    (bal/call "bal/aa"
              (list
               (cons "Account" "string")
               (cons "Type" "type")
               (cons "Opening Balance" "real")))))

(define at
  (lambda ()
    (bal/call "bal/at"
              (list
               (cons "Account" "current_account")
               (cons "Amount" "real")
               (cons "Description" "string")
               (cons "Day" "day")))))



(define ltn
  (lambda ()
    (let ((tscts (bal/call "bal/get-transactions"
                           (list
                            (cons "Account" "current_account")
                            (cons "How many?" "integer")))))
      (bal/print-tscts tscts))))


(define et
  (lambda* (#:optional n)
    (if n (bal/set-select-transact-num n))
    (bal/call "bal/edit-transact"
              (list
               (cons "Transaction" "transaction")
               (cons "Day (default is current day)" "day")
               (cons "Amount" "string")
               (cons "Description" "string")))))

(define lt
  (lambda ()
    (let* ((tscts (bal/get-transactions-by-day
                   (bal/get-current-account)
                   (list 0 0 0)
                   (bal/get-current-day)))
           (final-tsct (list-tail tscts
                                  (max (- (length tscts)
                                          bal/number-to-quick-list) 0))))
      (if (and (list? final-tsct)
               (not (null? final-tsct)))
          (bal/print-tscts final-tsct)))))

(define ea
  (lambda ()
    (bal/call "bal/ea"
              (list
               (cons "Account" "account")
               (cons "New account name" "string")
               (cons "Opening balance" "string")))))

(define da
  (lambda ()
    (bal/call "bal/da"
              (list
               (cons "Account" "account")))))

(define dt
  (lambda ()
    (bal/call "bal/dt"
              (list
               (cons "Transaction" "transaction")))))

(define la
  (lambda ()
    (bal/display-account-totals (bal/total-all-accounts))))

(define lae
  (lambda ()
    (bal/display-account-totals
     (bal/total-all-accounts-of-type 0))))

(define lai
  (lambda ()
    (bal/display-account-totals
     (bal/total-all-accounts-of-type 1))))

(define laa
  (lambda ()
    (bal/display-account-totals
     (bal/total-all-accounts-of-type 2))))

(define lal
  (lambda ()
    (bal/display-account-totals
     (bal/total-all-accounts-of-type 3))))


(define bt
  (lambda ()
    (let* ((accts (bal/total-by-account-type))
           (len
            (apply max
                   (map
                    (lambda (x) (string-length (car x))) accts))))
      (map-in-order
       (lambda (x)
         (display
          (if (pair? (cdr x))
              (string-append
               (format #f
                       (string-append "~"
                                      (number->string (+ len 5))
                                      "a") (car x))
               " "
               (format #f "~10,2f" (car (cdr x)))
               " "
               (format #f "~10,2f" (cdr (cdr x)))
               "\n")
              (string-append
               (format #f
                       (string-append "~"
                                      (number->string (+ len 5))
                                      "a") (car x))
               " "
               (format #f "~10,2f" (cdr x))
               "\n"))))
       accts))))

(define cex (bal/current-total-of-type 0))
(define cin (bal/current-total-of-type 1))
(define cas (bal/current-total-of-type 2))
(define cli (bal/current-total-of-type 3))
(define cwo (bal/current-total-of-type 4))
(define cba (bal/current-total-of-type 5))

(define re
  (lambda ()
    (bal/print-tscts
     (bal/call "bal/get-transactions-by-regex"
               (list
                (cons "Account" "current_account")
                (cons "Regular Expression" "string"))))))

(define sa
  (lambda ()
    (bal/call "bal/set-account"
              (list
               (cons "Account" "account")))))

(define ca
  (lambda ()
    (display
     (string-append (bal/get-current-account) "\n"))))

(define w
  (lambda ()
    (bal/call "bal/write"
              (list
               (cons "File" "string")))))

(define r
  (lambda ()
    (bal/call "bal/read"
              (list
               (cons "File" "string")))))

(define t
  (lambda ()
    (bal/call "bal/t"
              (list
               (cons "To Account" "account")
               (cons "From Account" "account")
               (cons "Amount" "real")
               (cons "Description" "string")
               (cons "Day" "day")))))

(define dtr
  (lambda ()
    (bal/call "bal/dtr"
              (list
               (cons "From Account" "account")
               (cons "Transaction" "transaction")))))


(define ltbd
  (lambda ()
    (bal/print-tscts
     (bal/call "bal/get-transactions-by-day"
               (list
                (cons "Account" "current_account")
                (cons "From Day" "day")
                (cons "To Day" "day"))))))

(define v
  (lambda ()
    (display
     (string-append
      "bal version: "
      (bal/v)
      "\n"))))

(define sd
  (lambda ()
    (bal/call "bal/set-current-day"
              (list
               (cons "Current Day" "day")))))

(define cd
  (lambda ()
    (let ((current-day (bal/get-current-day)))
      (display (string-append
                (number->string (list-ref current-day 2))
                "-"
                (format #f "~2,'0d" (list-ref current-day 1))
                "-"
                (format #f "~2,'0d" (list-ref current-day 0))
                "\n")))))

(define baod
  (lambda ()
    (let ((result (bal/call "bal/balance-account-on-days"
                            (list
                             (cons "First Day" "day")
                             (cons "Last Day" "day")
                             (cons "By" "number")
                             (cons "Account" "current_account")))))
      (map-in-order
       (lambda (x)
         (bal/output-by-day (car x) (cdr x)))
       result))))

(define exod
  (lambda ()
    (bal/over-day-cmd 0)))

(define inod
  (lambda ()
    (bal/over-day-cmd 1)))

(define asod
  (lambda ()
    (bal/over-day-cmd 2)))

(define liod
  (lambda ()
    (bal/over-day-cmd 3)))

(define wood
  (lambda ()
    (bal/over-day-cmd 4)))

(define ttbd
  (lambda ()
    (let ((result (bal/call
                   "bal/total-transact-in-account-between-days"
                   (list
                    (cons "First day" "day")
                    (cons "Last day" "day")
                    (cons "By" "number")
                    (cons "Account" "current_account")))))
      (map-in-order
       (lambda (x)
         (bal/output-by-day (car x) (cdr x)))
       result))))

(define ttre
  (lambda ()
    (let ((result (bal/call
                   "bal/total-transact-in-account-re"
                   (list
                    (cons "First day" "day")
                    (cons "Last day" "day")
                    (cons "By" "number")
                    (cons "Account" "current_account")
                    (cons "Regex" "string")))))
      (map-in-order
       (lambda (x)
         (bal/output-by-day (car x) (cdr x)))
       result))))


(define pl
  (lambda ()
    (bal/call "bal/pay-loan"
              (list
               (cons "Loan Account" "account")
               (cons "Interest Account" "account")
               (cons "Pay from Account" "account")
               (cons "Principal" "real")
               (cons "Interest" "real")
               (cons "Description" "string")
               (cons "Day" "day")))))

(define fn
  (lambda ()
    (display
     (string-append (bal/get-current-file) "\n"))))


(define csp
  (lambda ()
    (bal/call "bal/change-stock-price"
              (list
               (cons "To Account" "current_account")
               (cons "From Account" "account")
               (cons "Stock Price" "real")
               (cons "Number of Shares" "real")
               (cons "Day" "day")))))

(define cal
  (lambda* (#:optional args)
    (let ((cmd (string-append "cal "
                              (if args args ""))))
      (system cmd))))

